src/fat.rs

```diff
@@ -964,33 +949,120 @@ impl FatVolume {
                 while let Some(cluster) = current_cluster {
                     let block_idx = self.cluster_to_block(cluster);
                     for block in block_idx.range(BlockCount(u32::from(self.blocks_per_cluster))) {
-                        controller
-                            .block_device
-                            .read(&mut blocks, block, "read_dir")
-                            .map_err(Error::DeviceError)?;
-                        for entry in 0..Block::LEN / OnDiskDirEntry::LEN {
-                            let start = entry * OnDiskDirEntry::LEN;
-                            let end = (entry + 1) * OnDiskDirEntry::LEN;
-                            let dir_entry = OnDiskDirEntry::new(&blocks[0][start..end]);
-                            if dir_entry.is_end() {
-                                // Can quit early
-                                return Err(Error::FileNotFound);
-                            } else if dir_entry.matches(&match_name) {
-                                // Found it
-                                // Safe, since Block::LEN always fits on a u32
-                                let start = u32::try_from(start).unwrap();
-                                return Ok(dir_entry.get_entry(FatType::Fat16, block, start));
+                        match self.find_entry_in_block(controller, &match_name, block) {
+                            Err(Error::NotInBlock) => continue,
+                            x => return x,
+                        }
+                    }
+                    current_cluster = match self.next_cluster(controller, cluster) {
+                        Ok(n) => Some(n),
+                        _ => None,
+                    }
+                }
+                Err(Error::FileNotFound)
+            }
+        }
+    }
+
+    /// Finds an entry in a given block
+    fn find_entry_in_block<D, T>(
+        &self,
+        controller: &mut Controller<D, T>,
+        match_name: &ShortFileName,
+        block: BlockIdx,
+    ) -> Result<DirEntry, Error<D::Error>>
+    where
+        D: BlockDevice,
+        T: TimeSource,
+    {
+        let mut blocks = [Block::new()];
+        controller
+            .block_device
+            .read(&mut blocks, block, "read_dir")
+            .map_err(Error::DeviceError)?;
+        for entry in 0..Block::LEN / OnDiskDirEntry::LEN {
+            let start = entry * OnDiskDirEntry::LEN;
+            let end = (entry + 1) * OnDiskDirEntry::LEN;
+            let dir_entry = OnDiskDirEntry::new(&blocks[0][start..end]);
+            if dir_entry.is_end() {
+                // Can quit early
+                return Err(Error::FileNotFound);
+            } else if dir_entry.matches(&match_name) {
+                // Found it
+                // Safe, since Block::LEN always fits on a u32
+                let start = u32::try_from(start).unwrap();
+                return Ok(dir_entry.get_entry(FatType::Fat16, block, start));
```

Comment:
This seems to be a bug on the old code where it uses `FatType::Fat16` for both FAT16 and FAT32.

Reply:
Thanks for the input! I'll take a look at it, but I tested it with Fat16 and Fat32 and did not see any differences.
I'll take a look at it later tonight

Reply:
Fixed it. Now the proper FatType is being used for Fat32 entries